File: java/src/main/java/com/voteU/election/java/services/CountriesService.java
package com.voteU.election.java.services;

import com.voteU.election.java.entities.Countries;
import com.voteU.election.java.repositories.CountriesRepository;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;


/**
 * Service class for managing Countries entities. This service provides methods
 * to perform operations such as retrieving, searching, and accessing country data
 * through interactions with the CountriesRepository.
 */
@Service
public class CountriesService {

    private final CountriesRepository countriesRepository;

    public CountriesService(CountriesRepository countriesRepository) {
        this.countriesRepository = countriesRepository;
    }

    /**
     * Retrieves a list of all countries from the database.
     *
     * @return a list of Countries objects representing all countries available in the system.
     */
    public List<Countries> getAllCountries(){
        return countriesRepository.findAll();
    }

    /**
     * Retrieves a country entity based on the provided unique identifier.
     *
     * @param id the unique identifier of the country to retrieve
     * @return an Optional containing the Countries entity if found, or an empty Optional if no country exists with the specified id
     */
    public Optional<Countries> getCountryById(Integer id){
        return countriesRepository.findCountryById(id);
    }

    /**
     * Retrieves a country from the database based on the provided country code.
     *
     * @param code the unique code of the country to retrieve
     * @return an Optional containing the Countries entity if found, or an empty Optional if no country exists with the specified code
     */
    public Optional<Countries> getCountryByCode(String code){
        return countriesRepository.findCountryByCode(code);
    }

    /**
     * Retrieves a country entity from the database based on the provided name.
     *
     * @param name the name of the country to retrieve
     * @return an Optional containing the Countries entity if found, or an empty Optional if no country exists with the specified name
     */
    public Optional<Countries> getCountryByName(String name) {
        return countriesRepository.findCountryByName(name);
    }

}

File: voteu-project/src/services/CountriesService.ts
import { apiFetch } from '@/services/api.ts'
import type { Countries } from '@/interface/Countries.ts'

/**
 * Fetches and returns a list of all countries from the server.
 *
 * @return {Promise<Countries[]>} A promise that resolves to an array of country objects.
 */
export async function getAllCountries(): Promise<Countries[]> {
  return await apiFetch<Countries[]>('/countries',
    {
      method: 'GET',
    });
}

/**
 * Fetches and returns a country by its unique identifier.
 *
 * @param {number} id - The unique identifier of the country to be fetched.
 * @return {Promise<Countries>} A promise that resolves to the country data.
 */
export async function getCountryById(id: number): Promise<Countries> {
  return await apiFetch<Countries>(`/countries/${id}`);
}

/**
 * Retrieves a country's data by its code.
 *
 * @param {string} code - The unique code representing a country (e.g., ISO alpha-2 or alpha-3 code).
 * @return {Promise<Countries>} A promise that resolves to the country details matching the provided code.
 */
export async function getCountryByCode(code: string): Promise<Countries> {
  return await apiFetch<Countries>(`/countries/code/${code}`);
}

/**
 * Fetches country information by its name from the API.
 *
 * @param {string} name - The name of the country to retrieve.
 * @return {Promise<Countries>} A promise that resolves with the country data.
 */
export async function getCountryByName(name: string): Promise<Countries> {
  return await apiFetch<Countries>(`/countries/name/${name}`);
}
